src/transformers/models/vibevoice/generation_vibevoice.py

```diff
@@ -0,0 +1,653 @@
+# coding=utf-8
+# Copyright 2025 The Microsoft Team and The HuggingFace Inc. team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from queue import Queue
+from typing import Optional, Union
+
+import torch
+import torch.nn as nn
+from tqdm import tqdm
+
+from ...generation import (
+    BaseStreamer,
+    GenerateDecoderOnlyOutput,
+    GenerationConfig,
+    GenerationMixin,
+    LogitsProcessor,
+    LogitsProcessorList,
+)
+from ...utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+
+@dataclass
+class VibeVoiceGenerateOutput(GenerateDecoderOnlyOutput):
+    """
+    Outputs of VibeVoiceForConditionalGeneration.generate.
+
+    Args:
+        sequences (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
+            The generated sequences. The second dimension (sequence_length) is either equal to `max_length` or shorter
+            if all batches finished early due to the `eos_token_id`.
+        speech_outputs (`List[torch.FloatTensor]`, *optional*):
+            List of generated speech waveforms for each speech segment.
+        reach_max_step_sample (`torch.BoolTensor`, *optional*):
+            Boolean tensor indicating which samples reached maximum generation steps.
+    """
+    speech_outputs: Optional[list[torch.FloatTensor]] = None
+    reach_max_step_sample: Optional[torch.BoolTensor] = None
+
+
+class VibeVoiceTokenConstraintProcessor(LogitsProcessor):
+    """Constrains token generation to only valid tokens during speech generation."""
+
+    def __init__(self, valid_token_ids: list[int], device: torch.device = None):
+        self.valid_token_ids = torch.tensor(valid_token_ids, dtype=torch.long, device=device)
+
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
+        # Create a mask for valid tokens
+        mask = torch.full_like(scores, float('-inf'))
+        mask[:, self.valid_token_ids] = 0
+
+        # Apply mask to scores
+        scores = scores + mask
+        return scores
+
+
+# TODO (ebezzam) used in Gradio demo: https://github.com/vibevoice-community/VibeVoice/blob/main/demo/gradio_demo.py
+class AudioStreamer(BaseStreamer):
```

Comment:
There is currently no `AudioStreamer` within Transformers: https://github.com/huggingface/transformers/blob/main/src/transformers/generation/streamers.py

So this could be a nice feature to the library itself, but with @eustlb we'd rather keep it model-specific for now to think how it can be more model-agnostic / see how others would use such components

Reply:
Lmk if you want to take a look, happy to help on this!